cmd/traffic-controller/main_test.go

```diff
@@ -0,0 +1,294 @@
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+	"sigs.k8s.io/e2e-framework/pkg/envfuncs"
+	"sigs.k8s.io/e2e-framework/support/kind"
+	"sigs.k8s.io/e2e-framework/third_party/helm"
+	"sigs.k8s.io/external-dns/endpoint"
+)
+
+var (
+	testenv             env.Environment
+	releaseName         = "k8s-traffic-controller"
+	kindClusterName     = envconf.RandomName("traffic-controller", 16)
+	controllerNamespace = envconf.RandomName("controller", 16)
+	testNamespace       = envconf.RandomName("ingress", 16)
+)
+
+func deleteControllerDeployment(t *testing.T, k8sClient client.Client) {
+	t.Helper()
+
+	require.NoError(t, k8sClient.Delete(context.Background(), &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      releaseName + "-controller-manager",
+			Namespace: controllerNamespace,
+		},
+	}))
+}
+
+func installControllerChart(ctx context.Context, t *testing.T, k8sClient client.Client, args ...string) {
+	t.Helper()
+	helmClient := helm.New(testenv.EnvConf().KubeconfigFile())
+
+	require.NoError(t, helmClient.RunUpgrade(
+		helm.WithName(releaseName),
+		helm.WithNamespace(controllerNamespace),
+		helm.WithChart("../../helm-chart/traffic-controller"),
+		helm.WithArgs(
+			"--install",
+		),
+		helm.WithArgs(args...),
+	))
+}
+
+func installCRD(ctx context.Context, t *testing.T, k8sClient client.Client, url string) {
+	t.Helper()
+
+	resp, err := http.Get(url)
+	require.NoError(t, err)
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	decoder := yaml.NewYAMLOrJSONDecoder(resp.Body, 4096)
+
+	crd := &apiextensionsv1.CustomResourceDefinition{}
+	require.NoError(t, decoder.Decode(crd))
+
+	require.NoError(t, k8sClient.Create(ctx, crd))
+}
+
+func startMain(t *testing.T, args ...string) {
+	t.Helper()
+	os.Args = args
+	go func() {
+		main()
+	}()
+}
+
+func TestTrafficControllerController(t *testing.T) {
+	testutils.IntegrationTest(t)
+
+	t.Setenv("KUBECONFIG", testenv.EnvConf().KubeconfigFile())
+	osArgs := os.Args
+	originalContext := mainContext
+	t.Cleanup(func() {
+		mainContext = originalContext
+		os.Args = osArgs
+	})
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	mainContext = ctx
+
+	cfg, err := k8s.NewClientConfigBuilder().WithKubeConfigPath(testenv.EnvConf().KubeconfigFile()).Build()
+	require.NoError(t, err)
+
+	require.NoError(t, err)
+	k8sClient, err := client.New(cfg, client.Options{Scheme: scheme})
+	require.NoError(t, err)
+
+	installCRD(ctx, t, k8sClient, "https://raw.githubusercontent.com/kubernetes-sigs/external-dns/refs/heads/master/docs/contributing/crd-source/crd-manifest.yaml")
+	installControllerChart(ctx, t, k8sClient, "--set", "devMode=true")
+	deleteControllerDeployment(t, k8sClient)
+
+	sa := &v1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      releaseName,
+			Namespace: controllerNamespace,
+		},
+	}
+	assert.Eventually(t, func() bool {
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(sa), sa)
+		return err == nil
+	}, 5*time.Minute, 5*time.Second)
```

Comment:
my only comments is that 5m seems like a long time waiting for a SA to be created